### PR TITLE
fixed black background of transparent png thumbnail for older gd version

### LINKED ISSUE #236: Black background on resize/rotate for transparent images
The issue is the same as  #192 and #188, which are closed, but not fixed. I have the same problem with transparent png files. `bestFit ` works, but `thumbnail` fails on `crop`.

**Sample code:**

```php
$img = new \claviska\SimpleImage();
$img
  ->fromFile('png-32.png')
  ->thumbnail(100, 80)
  ->toFile('thumbnail.png');
```

**setup:**

SimpleImage: latest (3.3.3)

CentOS 7

PHP version: PHP 7.2.12 (cli) (built: Nov  6 2018 15:26:22) ( NTS )
Zend Engine: v3.2.0

GD library Version: 2.2.5
FreeType Version: 2.4.11
libJPEG Version: 6b
libPNG Version: 1.5.13 

**result:**

thumbnail

**original image:**

png-32

**local setup works:**

It works like expected on my localhost with Xampp, PHP 7.2 and

GD Version: bundled (2.1.0 compatible)
FreeType Version: 2.8.1
libJPEG Version: 9 compatible
libPNG Version: 1.6.34

**fix:**

I found a code snippet in SimpleImage for transparent gif files, modified it a bit and added it to the `crop` method:

https://github.com/raffaelj/SimpleImage/commit/31959f0f38f5b91d67e1bf47ea9a0deb96ac213c

**What I really want to do:**

Oh - and the actual reason is, to fix black thumbnails in [Cockpit CMS](https://github.com/agentejo/cockpit/) on the shared host [Uberspace](https://manual.uberspace.de/en/), where I can't change the gd version. Their support said

> PHP is compiled against the latest version and it's not possible to change it easily.

### PR DESCRIPTION
**Pull request by raffaelj**
fixes #236

### CODE REVIEW

File: src/claviska/SimpleImage.php [hunk 1]
```diff
@@ -579,15 +578,47 @@ public function crop($x1, $y1, $x2, $y2) {
 
       // Swap out the new image
       $this->image = $newImage;
-      break;
-    default:
+    }
+    elseif ($this->mimeType == 'image/gif' && GD_BUNDLED === 0){
```
Comment (claviska) [inline src/claviska/SimpleImage.php]: Same line as the previous closing bracket, for consistency :)

File: src/claviska/SimpleImage.php [hunk 2]
```diff
@@ -579,15 +578,47 @@ public function crop($x1, $y1, $x2, $y2) {
...
+        
+        $newImage = imagecreatetruecolor($width, $height);
+        
+        $trnprt_indx = imagecolortransparent($this->image);
```
Comment (claviska) [inline src/claviska/SimpleImage.php]: Please use camelCase vars for consistency.

File: src/claviska/SimpleImage.php [hunk 3]
```diff
@@ -579,15 +578,47 @@ public function crop($x1, $y1, $x2, $y2) {
...
+          0, 0, $x, $y,
+          $width,
+          $height,
+          $width,
+          $height
+        );
+        
+        $this->image = $newImage;
+    }
+    else {
```
Comment (claviska) [inline src/claviska/SimpleImage.php]: Same line as bracket, for consistency :)

File: src/claviska/SimpleImage.php [hunk 4]
```diff
@@ -842,12 +910,44 @@ public function rotate($angle, $backgroundColor = 'transparent') {
...
 
-    $this->image = imagerotate(
-      $this->image,
-      -(self::keepWithin($angle, -360, 360)),
-      $backgroundColor
-    );
-    imagecolortransparent($this->image, imagecolorallocatealpha($this->image, 0, 0, 0, 127));
+    if ($this->mimeType == 'image/gif' && GD_BUNDLED === 1 && $this->bits == 8){
+        
+      $trnprt_indx = imagecolortransparent($this->image);
```
Comment (claviska) [inline src/claviska/SimpleImage.php]: Use camelCase vars

File: src/claviska/SimpleImage.php [hunk 5]
```diff
@@ -842,12 +910,44 @@ public function rotate($angle, $backgroundColor = 'transparent') {
...
+      imagecolortransparent($this->image, $trnprt_indx);
+      
+      imagefill($this->image, 0, 0, $transparentColor);
+      imagecolortransparent($this->image, $transparentColor);
+      
+      imagefill($this->image, 0, 0, $backgroundColor);
+      
+    }
+    
+    else {
```
Comment (claviska) [inline src/claviska/SimpleImage.php]: Same line

**Comment by claviska** [comment]
I really appreciate your work on this. I suggested a couple of minor corrections to the code (the project doesn't have a linter config yet).

Since I have to maintain this longterm, I'd really like to see some comments detailing the new conditionals. For example, why do we check for `GD_BUNDLE`? And why do we care how many bits an image uses?

Feel free to be verbose if you need to explain it better (links are OK but could break in the future). This will ensure future PRs don't undo your work and help all maintainers understand what's going on.

Thanks again!

**Comment by raffaelj** [comment]
I did the cleanup and added some comments. I can't explain everything. Understanding all these gd functions would require a much bigger test setup with single functions and tons of test images. But I hope, I explained it good enough.

There was another stupid, but small bug in my code:

`imagescale($this->image, $width, $height);` must be `$this->image = imagescale($this->image, $width, $height);`

In the screenshots the jpeg files didn't resize anymore. Now it's fixed.

I moved the part for the rotate function into [a new branch](https://github.com/raffaelj/SimpleImage/tree/rotate-transp-gif-bundled). I'm close... but some images work, others are half-filled with artefacts... I won't repair this function. If I ever want to rotate transparent gif files with PHP in non 90° steps, I'll have a look at ImageMagick or I'll simply use CSS to rotate these damn things in the browser.

**Comment by raffaelj** [comment]
Don't merge it. I found some more issues. My current fix loses transparency in black image parts of gifs.

**Comment by adaniello** [comment]
@raffaelj no news about your PR? Thanks

**Comment by raffaelj** [comment]
@adaniello Sorry, I gave up on this. It took me ages to understand the most things and each approach broke another thing. This was my conclusion:
https://github.com/claviska/SimpleImage/issues/236#issuecomment-443920493

And a quote from claviska:

> GD can definitely be very frustrating, which is why I wrote SimpleImage in the first place. This issue, however, has proven impossible to perfect — thanks for trying and sharing your experience!

However, I am using my modified fork on some production setups and it seems to work as long as you don't try to apply filters on transparent images. This is an example, how I used the modified version in Cockpit CMS:
https://github.com/raffaelj/cockpit_SimpleImageFixBlackBackgrounds

**Comment by adaniello** [comment]
@raffaelj Thanks for your feedback and your time

**Comment by claviska** [comment]
Closing due to lack of activity.

**Pull request closed without merging**